protocol/node.go

```diff
@@ -2,52 +2,91 @@ package protocol
 
 import (
 	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
 	"github.com/monnand/dhkx"
 	"github.com/perlin-network/noise/log"
 	"github.com/pkg/errors"
 	"sync"
+	"sync/atomic"
+	"time"
 )
 
-type Service func(message *Message)
+var (
+	_ SendAdapter = (*Node)(nil)
+)
 
+// Node is a struct that wraps all the send/receive handlers
 type Node struct {
-	controller               *Controller
-	connAdapter              ConnectionAdapter
-	idAdapter                IdentityAdapter
-	peers                    sync.Map // string -> *PendingPeer | *EstablishedPeer
-	services                 map[uint16]Service
+	controller  *Controller
+	connAdapter ConnectionAdapter
+	idAdapter   IdentityAdapter
+
+	services                 []ServiceInterface
 	dhGroup                  *dhkx.DHGroup
 	dhKeypair                *dhkx.DHKey
 	customHandshakeProcessor HandshakeProcessor
+
+	// string -> *PendingPeer | *EstablishedPeer
+	peers sync.Map
+
+	// uint64 -> *RequestState
+	Requests     sync.Map
+	RequestNonce uint64
 }
 
-func NewNode(c *Controller, ca ConnectionAdapter, id IdentityAdapter) *Node {
-	g, err := dhkx.GetGroup(0)
+// RequestState represents a state of a request.
+type RequestState struct {
+	data        chan *MessageBody
+	closeSignal chan struct{}
+	requestTime time.Time
+}
+
+// NewNode constructs a new instance of Node
+func NewNode(c *Controller, id IdentityAdapter) *Node {
+	dhGroup, err := dhkx.GetGroup(0)
 	if err != nil {
 		panic(err)
 	}
 
-	privKey, err := g.GeneratePrivateKey(nil)
+	dhKeypair, err := dhGroup.GeneratePrivateKey(nil)
 	if err != nil {
 		panic(err)
 	}
 
 	return &Node{
-		controller:  c,
-		connAdapter: ca,
-		idAdapter:   id,
-		services:    make(map[uint16]Service),
-		dhGroup:     g,
-		dhKeypair:   privKey,
+		controller:   c,
+		idAdapter:    id,
+		services:     []ServiceInterface{},
+		dhGroup:      dhGroup,
+		dhKeypair:    dhKeypair,
+		RequestNonce: 0,
 	}
 }
 
+func (n *Node) SetConnectionAdapter(ca ConnectionAdapter) {
+	n.connAdapter = ca
+}
+
 func (n *Node) SetCustomHandshakeProcessor(p HandshakeProcessor) {
 	n.customHandshakeProcessor = p
 }
 
-func (n *Node) AddService(id uint16, s Service) {
-	n.services[id] = s
+func (n *Node) AddService(s ServiceInterface) {
+	n.services = append(n.services, s)
+}
+
+func (n *Node) GetIdentityAdapter() IdentityAdapter {
+	return n.idAdapter
+}
+
+func (n *Node) GetConnectionAdapter() ConnectionAdapter {
+	return n.connAdapter
+}
+
+func (n *Node) ManuallyRemovePeer(remote []byte) {
```

Comment:
why don't we just make removePeer RemovePeer?

Reply:
I'll remove it